creator-node/src/routes/users.js

```diff
@@ -41,16 +49,74 @@ module.exports = function (app) {
     }
 
-    // signature data is valid and matches a user, proceed with verifying timestamp is
-    // within time boundary
     const theirTimestamp = parseInt(data.split(':')[1])
     const ourTimestamp = Math.round((new Date()).getTime() / 1000)
 
     if (Math.abs(theirTimestamp - ourTimestamp) > 3600) {
       console.error(`Timestamp too old. User timestamp ${theirTimestamp}, Server timestamp ${ourTimestamp}`)
     }
 
-    // all checks have passed! generate a new session token for the user
+    const sessionToken = await sessionManager.createSession(user.cnodeUserUUID)
+    return successResponse({ sessionToken })
+  }))
+
+  /**
+   * Return a challeng used for validating user login. Challenge value
+   * is also set in redis cache with the key 'userLoginChallenge:<wallet>'.
+   */
+  app.get('/users/login/challenge', handleResponse(async (req, res, next) => {
+    const walletPublicKey = req.query.walletPublicKey
+    const userLoginChallengeKey = `userLoginChallenge:${walletPublicKey}`
+    const redisClient = req.app.get('redisClient')
+    const challengeBuffer = await randomBytes(CHALLENGE_VALUE_LENGTH)
+    const challenge = base64url.encode(challengeBuffer)
+
+    // Set challenge ttl to 10 minutes ('EX' option = sets expire time in seconds)
+    // https://redis.io/commands/set
+    // https://github.com/luin/ioredis/blob/master/examples/basic_operations.js#L44
+    await redisClient.set(userLoginChallengeKey, challenge, 'EX', CHALLENGE_TTL_SECONDS)
+
+    return successResponse({ walletPublicKey, challenge })
+  }))
+
+  /**
+   * Checks if challenge in request body matches up with what we have stored.
+   * If request challenge matches what we have, remove instance from redis to
+   * prevent replay attacks. Return sessionToken upon success.
+   */
+  app.post('/users/login/challenge', handleResponse(async (req, res, next) => {
+    const { signature, data, challenge } = req.body
+
+    const address = utils.verifySignature(data, signature)
+    const user = await models.CNodeUser.findOne({
+      where: {
+        walletPublicKey: address
+      }
+    })
+    if (!user) {
+      return errorResponseBadRequest('Invalid data or signature')
+    }
+
+    const theirChallengeBytes = data.split(': ')[1]
+
+    req.logger.info(`Detected challenge ${challenge}`)
+    const redisClient = req.app.get('redisClient')
+    const userLoginChallengeKey = `userLoginChallenge:${address}`
+    const ourChallengeBytes = await redisClient.get(userLoginChallengeKey)
+
+    if (!ourChallengeBytes) {
+      return errorResponseBadRequest('Missing challenge key')
+    }
+
+    if (theirChallengeBytes !== ourChallengeBytes) {
+      return errorResponseBadRequest(`Found invalid response: User Challenge Value ${theirChallengeBytes}, Server Challenge Value ${ourChallengeBytes}`)
+    } else {
+      req.logger.info(`Found equal signatures: User Challenge Value ${theirChallengeBytes}, Server Challenge Value ${ourChallengeBytes}`)
```

Comment:
remove this log or don't log the challenge bytes. this could be a huge security leak if you give them `ourChallengeBytes`